File: qtiscoringengineTester/src/main/java/qtiscoringengineTester/ScoreCounter.java
package qtiscoringengineTester;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

enum LogType {
  correct, incorrect, error
};

public class ScoreCounter
{
  private static final Logger                 _logger         = LoggerFactory.getLogger (ScoreCounter.class);
  private int                                 _incorrect      = 0;
  private int                                 _correct        = 0;
  private int                                 _exception      = 0;
  private int                                 _missingRubrics = 0;

  private Map<LogType, Map<Integer, Integer>> _histogram      = new HashMap<LogType, Map<Integer, Integer>> ();

  public synchronized void incrementIncorrect (int realScore) {
    ++_incorrect;
    addToMapForRealScore (realScore, LogType.incorrect);
    logTotal ();
  }

  public synchronized void incrementCorrect (int realScore) {
    ++_correct;
    addToMapForRealScore (realScore, LogType.correct);
    logTotal ();
  }

  public synchronized void incrementErrors (int realScore) {
    ++_exception;
    addToMapForRealScore (realScore, LogType.error);
    logTotal ();
  }

  public synchronized void incrementMissingRubrics () {
    ++_missingRubrics;
    logTotal ();
  }

  public int getIncorrect () {
    return _incorrect;
  }

  public int getCorrect () {
    return _correct;
  }

  public int getErrors () {
    return _exception;
  }

  private void logTotal () {
    // TODO Shiva: calculating everytime is inefficient.
    int total = getIncorrect () + getCorrect () + getErrors ();
    if (total % 2000 == 0) {
      _logger.error (this.toString ());
    }
  }

  @Override
  public String toString () {
    StringBuilder outputBuilder = new StringBuilder ();
    outputBuilder.append ("======================Stats====================\n");
    outputBuilder.append (String.format ("Correct: %d ; Incorrect: %d ; Errors : %d\n", this.getCorrect (), this.getIncorrect (), this.getErrors ()));
    outputBuilder.append (String.format ("Missing rubrics: %d\n", _missingRubrics));
    for (Map.Entry<LogType, Map<Integer, Integer>> entries : _histogram.entrySet ()) {
      LogType key = entries.getKey ();
      for (Map.Entry<Integer, Integer> record : entries.getValue ().entrySet ()) {
        outputBuilder.append (String.format ("Type: %s ; Score : %d ; Counter : %d\n", key.name (), record.getKey (), record.getValue ()));
      }
    }

    return outputBuilder.toString ();
  }

  /*
   * for type : correct, incorrect or error we want to keep track of the
   * distribution of scores.
   */
  private void addToMapForRealScore (int realScore, LogType type) {
    if (!_histogram.containsKey (type))
      _histogram.put (type, new HashMap<Integer, Integer> ());

    Map<Integer, Integer> scoreHistogram = _histogram.get (type);

    int scoreTracker = 0;
    if (scoreHistogram.containsKey (realScore))
      scoreTracker = scoreHistogram.get (realScore);

    scoreTracker++;
    scoreHistogram.put (realScore, scoreTracker);
  }
}

File: ASPNETCodeForComparisionTesting/QTITester/ScoreCounter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace QTITester
{

    enum LogType
    {
        correct, incorrect, error
    };

    public class ScoreCounter
    {
        private int _incorrect = 0;
        private int _correct = 0;
        private int _exception = 0;

        private Dictionary<LogType, Dictionary<int, int>> _histogram = new Dictionary<LogType, Dictionary<int, int>>();

        public void IncrementIncorrect(int realScore)
        {
            lock (this)
            {
                ++_incorrect;
                AddToMapForRealScore(realScore, LogType.incorrect);
            }
        }

        public void IncrementCorrect(int realScore)
        {
            lock (this)
            {
                ++_correct;
                AddToMapForRealScore(realScore, LogType.correct);
            }
        }

        public void IncrementErrors(int realScore)
        {
            lock (this)
            {
                ++_exception;
                AddToMapForRealScore(realScore, LogType.error);
            }
        }

        public int GetIncorrect()
        {
            return _incorrect;
        }

        public int GetCorrect()
        {
            return _correct;
        }

        public int GetErrors()
        {
            return _exception;
        }

        public override String ToString()
        {
            StringBuilder outputBuilder = new StringBuilder();
            outputBuilder.Append("======================Stats====================\n");
            outputBuilder.Append(String.Format("Correct: {0} ; Incorrect: {1} ; Errors : {2}\n", this.GetCorrect(), this.GetIncorrect(), this.GetErrors()));

            foreach (KeyValuePair<LogType, Dictionary<int, int>> entries in _histogram)
            {
                LogType key = entries.Key;
                foreach (KeyValuePair<int, int> record in entries.Value)
                {
                    outputBuilder.Append(String.Format("Type: {0} ; Score : {1} ; Counter : {2}\n", key, record.Key, record.Value));
                }
            }

            return outputBuilder.ToString();
        }

        /*
         * for type : correct, incorrect or error we want to keep track of the
         * distribution of scores.
         */
        private void AddToMapForRealScore(int realScore, LogType type)
        {
            if (!_histogram.ContainsKey(type))
                _histogram[type] = new Dictionary<int, int>();

            Dictionary<int, int> scoreHistogram = _histogram[type];

            int scoreTracker = 0;
            if (scoreHistogram.ContainsKey(realScore))
                scoreTracker = scoreHistogram[realScore];

            scoreTracker++;
            scoreHistogram[realScore] = scoreTracker;
        }
    }

}
